### PR TITLE
fix: Type check ACIR mutable reference passed to brillig

### LINKED ISSUE #4245: Passing a mut reference into an unconstrained function crashes the compiler
### Aim

Trying to work around #4244

Code:

```rust
use dep::std::field::bn254::assert_gt;

struct ListItem {
    key: Field,
    previous: Field,
    next: Field,
}

impl ListItem {
    fn default() -> ListItem {
        ListItem {
            key: 0,
            previous: 0,
            next: 0,
        }
    }
}

struct Map<Size> {
    entries: [ListItem; Size],
    size: Field,
    is_empty: bool,
    first_index: Field,
    last_index: Field,
}

unconstrained fn find_previous_key_location<Size>(map: Map<Size>, key: Field) -> (Field) {
    let mut found_index: Field = 0;
    found_index
}

unconstrained fn find_previous_key_location_nocopy<Size>(key: Field) -> (Field) {
    let mut found_index: Field = 0;
    found_index
}

unconstrained fn find_previous_key_location_crash<Size>(map: &mut Map<Size>, key: Field) -> (Field) {
    let mut found_index: Field = 0;
    found_index
}

impl<Size> Map<Size> {
    fn default() -> Map<Size> {
        Map{
            entries: [ListItem::default(); Size], // todo fix
            size: 0,
            is_empty: true,
            first_index: 0,
            last_index: 0,
        }
    }


    fn insert_crashy(&mut self, key: Field) {
        let previous_indexx = find_previous_key_location_crash(self, key);
        self.entries[previous_indexx].key = key;
    }

}

fn main(x: Field, y: pub Field) {
    let mut test_list: Map<50> = Map::default();
    test_list.insert_crashy(x);
    test_list.insert_crashy(x + y);
    test_list.insert_crashy(x + y * y);
}
```

### Expected Behavior

no crash. Sad face.

### Bug

compiler crashes

### To Reproduce

1. nargo info
2. observe crash
3.
4.

### Installation Method

None

### Nargo Version

0.23.0

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by vezenovm**
# Description

## Problem\*

Resolves #4245 

## Summary\*

This adds a type check for call expressions to check the arguments of a call to an unconstrained function from a constrained function. As we have two different runtime environments this makes type checking a bit more complicated. As a followup to this PR we could try and catch the error added here (https://github.com/noir-lang/noir/pull/4280) sooner. However, if the unconstrained function uses generics we must still fall back to the "type check" in SSA/ACIR gen from PR #4280.

This PR now errors for the snippet in issue with the below using `nargo check`:

Screenshot 2024-02-06 at 3 36 27 PM

## Documentation\*

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Review by jfecher — APPROVED** [review]
It'd be nice if we could get this check on methods too but this is an improvement over the current panic

File: test_programs/compile_failure/brillig_mut_ref_from_acir/src/main.nr
Comment (TomAFrench) [inline test_programs/compile_failure/brillig_mut_ref_from_acir/src/main.nr:1]: Can we reduce this example a bit? We don't need all these structs and impls, can we just pass a mutable reference to a single fields into a brillig identity function?

Comment (vezenovm) [reply test_programs/compile_failure/brillig_mut_ref_from_acir/src/main.nr:1]: Yeah fair enough a shorter example is better for debugging. Just pushed a much shorter example.

**Comment by vezenovm** [comment]
> It'd be nice if we could get this check on methods too but this is an improvement over the current panic

Good point, I can add this as a follow-up. For now I made this issue to keep track: https://github.com/noir-lang/noir/issues/4288

**Pull request merged**